### PR TITLE
Make Metadata use ContentType

### LINKED ISSUE #5538: Make `Metadata.content_type` use hyper's ContentType
[`Metadata`](http://doc.servo.org/net_traits/struct.Metadata.html) has a ContentType field, which is currently an `Option<(String, String)>`. We should probably be using an [`Option<ContentType>`](http://hyperium.github.io/hyper/hyper/header/struct.ContentType.html).

Relevant code: https://github.com/servo/servo/blob/master/components/net_traits/lib.rs#L107 (changing it will probably throw up a couple of compile errors elsewhere which will need to be updated too)

### PR DESCRIPTION
**Pull request by cuzbog**
Fixes #5538

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/5544)

### CODE REVIEW

File: components/script/parse/html.rs [hunk 1]
```diff
@@ -291,12 +293,12 @@ pub fn parse_html(document: JSRef<Document>,
...
         HTMLInput::InputUrl(load_response) => {
             match load_response.metadata.content_type {
-                Some((ref t, _)) if t.as_slice().eq_ignore_ascii_case("image") => {
+                Some(ContentType(Mime(ref t, _, _))) if *t == TopLevel::Image => {
                     let page = format!("<html><body><img src='{}' /></body></html>", url.serialize());
                     parser.parse_chunk(page);
                 },
-                Some((ref t, ref st)) if t.as_slice().eq_ignore_ascii_case("text") &&
-                                         st.as_slice().eq_ignore_ascii_case("plain") => {
+                Some(ContentType(Mime(ref t, ref st, _))) if *t == TopLevel::Text &&
```
Comment (boghison) [inline components/script/parse/html.rs]: Oh, yes, totally, I was just keeping the original structure :)

**Comment by jdm** [comment]
Sorry, looks like #5005 added some code that needs to be adjusted now.

```
/Users/servo/buildbot/slave/mac2/build/components/net/resource_task.rs:82:53: 82:75 error: mismatched types:
 expected `&core::option::Option<(collections::string::String, collections::string::String)>`,
    found `&core::option::Option<hyper::header::common::content_type::ContentType>`
(expected tuple,
    found struct `hyper::header::common::content_type::ContentType`) [E0308]
/Users/servo/buildbot/slave/mac2/build/components/net/resource_task.rs:82                                                     &metadata.content_type, &partial_body);
                                                                                                                              ^~~~~~~~~~~~~~~~~~~~~~
/Users/servo/buildbot/slave/mac2/build/components/net/resource_task.rs:81:33: 82:91 error: mismatched types:
 expected `core::option::Option<hyper::header::common::content_type::ContentType>`,
    found `core::option::Option<(collections::string::String, collections::string::String)>`
(expected struct `hyper::header::common::content_type::ContentType`,
    found tuple) [E0308]
/Users/servo/buildbot/slave/mac2/build/components/net/resource_task.rs:81         metadata.content_type = classifier.classify(nosniff, check_for_apache_bug,
/Users/servo/buildbot/slave/mac2/build/components/net/resource_task.rs:82                                                     &metadata.content_type, &partial_body);
error: aborting due to 2 previous errors
```

**Comment by boghison** [comment]
@jdm so what am I supposed to do, edit the whole mime classifier or just send toplevel and sublevel as strings?

**Comment by jdm** [comment]
I would go ahead and edit the classifier code.

**Comment by boghison** [comment]
@jdm That is going to need a lot of rewriting, mostly because almost everything has been fixed to use strings there

**Comment by jdm** [comment]
Oh, hmm. Go ahead and pass it strings, in that case. We can file a follow-up to rewrite it separately.

**Comment by boghison** [comment]
@jdm Yes, that would be best, the hardest part is probably dealing with non standard types like application+xml

**Comment by boghison** [comment]
@jdm How do I actually convert a mime to a string?

**Comment by jdm** [comment]
You might need to resort to `format!("{}", mime)` for the full a/b format, or `format!("{}", mime.0)` for the subparts in particular.

**Comment by jdm** [comment]
```
<std macros>:5:10: 5:35 error: the trait `core::cmp::PartialEq<core::option::Option<(collections::string::String, collections::string::String)>>` is not implemented for the type `core::option::Option<hyper::header::common::content_type::ContentType>` [E0277]
<std macros>:5 if ! ( ( * left_val == * right_val ) && ( * right_val == * left_val ) ) {
                        ^~~~~~~~~~~~~~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
/home/servo/buildbot/slave/linux1/build/tests/unit/net/data_loader.rs:21:5: 21:64 note: expansion site
<std macros>:5:43: 5:68 error: the trait `core::cmp::PartialEq<core::option::Option<hyper::header::common::content_type::ContentType>>` is not implemented for the type `core::option::Option<(collections::string::String, collections::string::String)>` [E0277]
<std macros>:5 if ! ( ( * left_val == * right_val ) && ( * right_val == * left_val ) ) {
                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
/home/servo/buildbot/slave/linux1/build/tests/unit/net/data_loader.rs:21:5: 21:64 note: expansion site
```
You can reproduce this with `./mach test-unit`.

**Comment by boghison** [comment]
@jdm To fix this test I'd have to add hyper as an external dependency, is it ok?

**Comment by Manishearth** [comment]
Yes, to the unit tests crate.

**Comment by boghison** [comment]
@Manishearth Yes, obviously

**Pull request merged**